test/gateway/index.js

```diff
@@ -0,0 +1,191 @@
+const expect = chai.expect
+chai.use(dirtyChai)
+const API = require('../../src/http')
+const loadFixture = require('aegir/fixtures')
+const bigFile = loadFixture(__dirname, '../../node_modules/interface-ipfs-core/test/fixtures/15mb.random', 'ipfs')
+const directoryContent = {
+  'index.html': loadFixture(__dirname, './test-folder/index.html', 'ipfs'),
+  'nested-folder/hello.txt': loadFixture(__dirname, './test-folder/nested-folder/hello.txt', 'ipfs'),
+  'nested-folder/ipfs.txt': loadFixture(__dirname, './test-folder/nested-folder/ipfs.txt', 'ipfs'),
+  'nested-folder/nested.html': loadFixture(__dirname, './test-folder/nested-folder/nested.html', 'ipfs')
+}
+
+describe('HTTP Gateway', () => {
+  let http = {}
+  let gateway
+
+  before((done) => {
+    http.api = new API()
+
+    http.api.start(true, () => {
+      const content = (name) => ({
+        path: `test-folder/${name}`,
+        content: directoryContent[name]
+      })
+
+      const emptyDir = (name) => ({
+        path: `test-folder/${name}`
+      })
+
+      const expectedRootMultihash = 'QmbQD7EMEL1zeebwBsWEfA3ndgSS6F7S6iTuwuqasPgVRi'
+
+      const dirs = [
+        content('index.html'),
+        emptyDir('empty-folder'),
+        content('nested-folder/hello.txt'),
+        content('nested-folder/ipfs.txt'),
+        content('nested-folder/nested.html'),
+        emptyDir('nested-folder/empty')
+      ]
+
+      http.api.node.files.add(dirs, (err, res) => {
+        expect(err).to.not.exist()
+        const root = res[res.length - 1]
+
+        expect(root.path).to.equal('test-folder')
+        expect(root.hash).to.equal(expectedRootMultihash)
+        gateway = http.api.server.select('Gateway')
+        done()
+      })
+    })
+  })
+
+  after((done) => {
+    http.api.stop((err) => {
+      expect(err).to.not.exist()
+      done()
+    })
+  })
+
+  describe('## HTTP Gateway', () => {
+    it('returns 400 for request without argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('400 for request with invalid argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/invalid'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('valid hash', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/QmT78zSuBmuS4z925WZfrqQ1qHaJ56DQaTfyMUF7F8ff5o'
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.rawPayload).to.deep.equal(Buffer.from('hello world' + '\n'))
+        expect(res.payload).to.equal('hello world' + '\n')
+        done()
+      })
+    })
+
+    it('stream a large file', (done) => {
+      let bigFileHash = 'Qme79tX2bViL26vNjPsF3DP1R9rMKMvnPYJiKTTKPrXJjq'
```

Comment:
@ya7ya where does this hash/file come from? Doesn't seem it got added anywhere

Reply:
Trying to load it through the gateway ipfs.io, I get:
![image](https://user-images.githubusercontent.com/1211152/30147246-3340a5ac-9395-11e7-85ff-048dbfa5c535.png)


https://ipfs.io/ipfs/Qme79tX2bViL26vNjPsF3DP1R9rMKMvnPYJiKTTKPrXJjq

Reply:
@diasdavid thats from the 15mb random file in the fixtures in `interface-ipfs-core` , I just noticed I forgot to add it to `files.add` in the before section. So sorry about this.